File: Service/BindServiceInAnotherApp/app/src/main/java/com/example/syp/bindserviceinanotherapp/MainActivity.java
package com.example.syp.bindserviceinanotherapp;

import android.content.ComponentName;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.os.RemoteException;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import com.example.anotherapp.IOnRemoteCountChangeListener;
import com.example.anotherapp.IRemoteBinder;

/**
 * AIDL远程回调
 */
public class MainActivity extends AppCompatActivity implements View.OnClickListener, ServiceConnection {
    private Intent myServiceIntent;
    private IRemoteBinder binder = null;
    private TextView outputText;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        outputText = (TextView) findViewById(R.id.outputText);

        myServiceIntent = new Intent();
        myServiceIntent.setComponent(new ComponentName("com.example.anotherapp", "com.example.anotherapp.MyService"));

        findViewById(R.id.btnBindRemoteService).setOnClickListener(this);
        findViewById(R.id.btnUnbindRemoteService).setOnClickListener(this);
        findViewById(R.id.btnSetCountTo100).setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btnBindRemoteService:
                bindService(myServiceIntent, this, BIND_AUTO_CREATE);
                break;
            case R.id.btnUnbindRemoteService:
                binder = null;
                unbindService(this);
                break;
            case R.id.btnSetCountTo100:
                if (binder != null) {
                    try {
                        binder.setCount(100);
                    } catch (RemoteException e) {
                        e.printStackTrace();
                    }
                }
                break;
        }
    }

    //会实现上述第一个工程MyService类中的 onBind方法
    @Override
    public void onServiceConnected(ComponentName name, IBinder service) {
        binder = IRemoteBinder.Stub.asInterface(service);

        try {
            binder.addOnRemoteCountChangeListener(new IOnRemoteCountChangeListener.Stub() {
                @Override
                public void onChange(int count) throws RemoteException {
                    final int c = count;
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            outputText.setText("Count " + c);
                        }
                    });
                }
            });
        } catch (RemoteException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void onServiceDisconnected(ComponentName name) {

    }
}

File: Service/AIDLCallBack/app/src/main/java/com/example/syp/aidlcallback/MainActivity.kt
package com.example.syp.aidlcallback;

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.os.IBinder
import android.view.View
import android.widget.TextView
import com.example.anotherapp.ICountChangeListener
import com.example.anotherapp.IRemoteBinder

/**
 * 2.AIDL进程间通信
 */

open class MainActivity : AppCompatActivity(), View.OnClickListener, ServiceConnection {

    override fun onServiceDisconnected(name: ComponentName?) {}

    override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
        remoteBinder = IRemoteBinder.Stub.asInterface(service)
    }

    private var remoteServiceIntent: Intent? = null
    private var remoteBinder: IRemoteBinder? = null

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btnBindRemoteService -> {
                bindService(remoteServiceIntent, this, Context.BIND_AUTO_CREATE)
            }
            R.id.btnUnbindRemoteService -> {
                remoteBinder = null
                unbindService(this)
            }
            R.id.btnSetCountTo100 -> {
                remoteBinder?.count = 100
            }
            R.id.btnRegisterRemoteListener -> {
                remoteBinder?.registerCountChangeListener(countChangeListener)
            }
            R.id.btnUnregisterRemoteListener -> {
                remoteBinder?.unregisterCountChangeListener(countChangeListener)
            }
        }
    }

    inner class CountChangeListenerImpl : ICountChangeListener.Stub() {
        override fun onChange(count: Int) {
            runOnUiThread {
                outputText?.text = "Count " + count
            }
        }
    }

    private val countChangeListener = CountChangeListenerImpl()


    private var outputText: TextView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        outputText = findViewById(R.id.outputText) as TextView?
        remoteServiceIntent = Intent()

        // remoteServiceIntent!!.setComponent(ComponentName("com.example.anotherapp", "com.example.anotherapp.MyService"))
        remoteServiceIntent!!.component = ComponentName("com.example.anotherapp", "com.example.anotherapp.MyService")

        findViewById(R.id.btnBindRemoteService).setOnClickListener(this)
        findViewById(R.id.btnUnbindRemoteService).setOnClickListener(this)
        findViewById(R.id.btnSetCountTo100).setOnClickListener(this)
        findViewById(R.id.btnRegisterRemoteListener).setOnClickListener(this)
        findViewById(R.id.btnUnregisterRemoteListener).setOnClickListener(this)

    }

}
